io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -0,0 +1,1574 @@
+updateRacesInSimState thread SimState{ control, threads, races } =
+  traceRaces $
+  updateRaces (currentStep thread)
+              (threadBlocked thread)
+              control
+              (Map.keysSet (Map.filter (not . threadDone) threads))
+              races
+
+-- We take care that steps can only race against threads in their
+-- concurrent set. When this becomes empty, a step can be retired into
+-- the "complete" category, but only if there are some steps racing
+-- with it.
+updateRaces :: Step -> Bool -> ScheduleControl -> Set ThreadId -> Races -> Races
+updateRaces newStep@Step{ stepThreadId = tid, stepEffect = newEffect }
+            blocking
+            control
+            newConcurrent0
+            races@Races{ activeRaces } =
+  let -- a new step cannot race with any threads that it just woke up
+      newConcurrent = foldr Set.delete newConcurrent0 (effectWakeup newEffect)
+      new | isTestThreadId tid     = []  -- test threads do not race
+          | Set.null newConcurrent = []  -- cannot race with anything
+          | justBlocking           = []  -- no need to defer a blocking transaction
+          | otherwise              =
+              [StepInfo { stepInfoStep       = newStep,
+                          stepInfoControl    = control,
+                          stepInfoConcurrent = newConcurrent,
+                          stepInfoNonDep     = [],
+                          stepInfoRaces      = []
+                        }]
+      justBlocking = blocking && onlyReadEffect newEffect
+      updateActive =
+        [ -- if this step depends on the previous step, or is not concurrent,
+          -- then any threads that it wakes up become non-concurrent also.
+          let lessConcurrent = foldr Set.delete concurrent (effectWakeup newEffect) in
+          if tid `elem` concurrent then
+            let theseStepsRace = not (isTestThreadId tid) && racingSteps step newStep
+                happensBefore  = hbfStep (stepThreadId step) (stepStep step) (stepVClock newStep)
+                nondep' | happensBefore = nondep
+                        | otherwise     = newStep : nondep
+                -- We will only record the first race with each thread---reversing
+                -- the first race makes the next race detectable. Thus we remove a
+                -- thread from the concurrent set after the first race.
+                concurrent' | happensBefore  = Set.delete tid lessConcurrent
+                            | theseStepsRace = Set.delete tid concurrent
+                            | otherwise      = concurrent
+                -- Here we record discovered races.
+                -- We only record a new race if we are following the default schedule,
+                -- to avoid finding the same race in different parts of the search space.
+                stepRaces' | (control == ControlDefault ||
+                              control == ControlFollow [] []) &&
```

Comment:
I don't think so. advanceControl can certainly return ControlFollow [] []. It may be that it's equivalent to ControlDefault, though. But I think it can certainly arise here.

Reply:
With small modification we could make sure `advanceControl` never returns `ControlFollow [] []`.

Reply:
Yes... although right now advanceControl moves from ControlFollow [step] [] to ControlFollow [] [] to ControlDefault in two steps. You're suggesting short-circuiting that to one step. That could break things elsewhere, unless ControlFollow [] [] and ControlDefault are treated as equivalent everywhere. I think they should be, but I'm not completely certain.

So, it's a small change, yes, but a small change that I would follow with extensive testing to make sure nothing broke.